src/options.rs

```diff
@@ -35,70 +60,101 @@ impl Options {
         &self.setter_pattern
     }
 
-    pub fn setter_visibility(&self) -> Option<quote::Tokens> {
-        if self.setter_public {
-            Some(quote!(pub))
-        } else {
-            None
-        }
+    pub fn setter_visibility(&self) -> &syn::Visibility {
+        &self.setter_vis
     }
 
     pub fn setter_prefix(&self) -> &str {
         &self.setter_prefix
     }
+
+    pub fn field_name(&self) -> &str {
+        &self.field_name
+    }
 }
 
-impl<'a, T> From<T> for Options where
-    T: IntoIterator<Item=&'a syn::Attribute>
-{
-    fn from(attributes: T) -> Self {
+impl<'a> From<&'a syn::MacroInput> for StructOptions {
+    fn from(ast: &'a syn::MacroInput) -> Self {
         trace!("Parsing struct attributes.");
         let mut builder = OptionsBuilder::<StructMode>::default();
-        builder.parse_attributes(attributes);
+        builder.parse_attributes(&ast.attrs);
+
+        builder.mode.struct_name = ast.ident.as_ref().to_string();
+        builder.mode.struct_vis = Some(ast.vis.clone());
 
         builder.into()
     }
 }
 
-pub trait OptionsBuilderMode {}
+pub trait OptionsBuilderMode {
+    fn parse_builder_name(&mut self, lit: &syn::Lit);
+}
 
 #[derive(Default)]
-pub struct StructMode;
+pub struct StructMode{
+    builder_name: Option<String>,
+    builder_vis: Option<syn::Visibility>,
+    struct_name: String,
+    struct_vis: Option<syn::Visibility>,
+}
 
-impl OptionsBuilderMode for StructMode {}
+impl OptionsBuilderMode for StructMode {
+    fn parse_builder_name(&mut self, lit: &syn::Lit) {
+        trace!("Parsing builder name {:?}", lit);
+        let value = parse_lit_as_cooked_string(lit);
+        self.builder_name = Some(value.clone());
+    }
+}
 
 #[derive(Default)]
 pub struct FieldMode;
 
-impl OptionsBuilderMode for FieldMode {}
+impl OptionsBuilderMode for FieldMode {
+    fn parse_builder_name(&mut self, _: &syn::Lit) {
```

Comment:
I tend to name even unused bindings to easer grasp what kind of literal I get here. Maybe `_name`?